### ISSUE TITLE
Reference for passing arguments to backward

### ISSUE THREAD

**Issue by dumbPy** [Zeleni9/pytorch-wgan#2]
https://github.com/Zeleni9/pytorch-wgan/blob/21f929be859b8fabdd36e4d4cbc4411cca7944ec/models/wgan_gradient_penalty.py#L180

Would you provide some reference for what you did here? I have never passed an argument to loss.backward() and why do you pass one and mone.

**Comment by Zeleni9** [comment]
Hi Sufiyan, 

that was pytorch trick in pytorch 0.3 I think to implement constant gradients. Check this reddit thread [https://www.reddit.com/r/MachineLearning/comments/5qxoaz/r_170107875_wasserstein_gan/ <https://www.reddit.com/r/MachineLearning/comments/5qxoaz/r_170107875_wasserstein_gan/>] to find out more why since it is explained well. I will copy paste the main part or ctrl-f the word “backward” in the thread.

My implementation trick with AbsCriterion is just meant as a way of implementing constant gradients in Torch. By setting the target to be output +- 1, we guarantee that the gradient will be +- 1.
My initial reason for using +- 1 gradients was pretty much that the comments here say to. However, I was able to muster some intuition for it. If you look at Eq. 3 in the paper, you will notice that the maximum is reached when f_w(x) yields high values for real examples and low values for generated examples. This is exactly what we are working towards by pushing the critic up and down with +-1 gradients.



level 9
JaejunYoo <https://www.reddit.com/user/JaejunYoo/>
1 point
·
2 years ago
 <https://www.reddit.com/r/MachineLearning/comments/5qxoaz/r_170107875_wasserstein_gan/de97zlh/>
THANK YOU VERY MUCH. I got your point. So.. basically it is due to the details required for Torch implementation. I was confused because I was not familiar with Torch. Thank you for your quick reply again.


Best, 
Filip.

**Comment by dumbPy** [comment]
Thank You very much.. Wouldn't have been able to find that discussion without you.
It's still a bit blurry to me though. Tried to find a mathematical explanation to what passing gradient tensor to backward does really. Couldn't find any reference to it in official docs.

Tried some examples in terms of tensors. Seems like the input tensor is just multiplied to the calculated gradient. Leaving it here in case my future self would like to take a look again.

Thank You again.
![Screenshot from 2019-07-31 18-42-11](https://user-images.githubusercontent.com/34011516/62214958-ad5c1b00-b3c3-11e9-8ca2-0f435bee9576.png)
